### PR TITLE
protobuf.js JSON descriptor

### PR DESCRIPTION
**Pull request by SOVLOOKUP**
**new feature:**
make `@grpc/proto-loader` can parse JSON descriptors in [protobuf.js](https://github.com/protobufjs/protobuf.js#using-json-descriptors)

**code change:**
encapsulated `fromJSON` function
and its test

### CODE REVIEW

**Comment by merlinnot** [comment]
Amazing, I was looking for it just last Friday!

**Comment by murgatroid99** [comment]
What is the use case for this?

**Comment by SOVLOOKUP** [comment]
@murgatroid99 I am building a plugin-system based on grpc. I need a protobuf definition with both readability and transmission convenience. I think the [protobufjs official json definition](https://github.com/protobufjs/protobuf.js#using-json-descriptors) is very suitable , but grpc-proto-loader does not support parsing this format. So I implemented it.

**Comment by merlinnot** [comment]
> What is the use case for this?

Currently proto-loader can only load files from disk (it has to read them), which also makes it harder to use in environments for which bundling is preferred (e.g. using Rollup or Webpack) or which disallow disk access completely. It is the most convenient to just import a JSON file (`import * as proto from './my/path/to/proto.json'`) and then call `fromJSON(proto)`.

An example of an environment where bundling to a single file is necessary: Cloudflare Workers.

**Comment by merlinnot** [comment]
Also, using it with ES modules is slightly inconvenient:

```ts
const packageDefinition = loadSync(
  new URL(join(dirname(import.meta.url), '../lib/bundle.json')).pathname,
);
```

as there is no `__dirname`.

File: packages/proto-loader/src/index.ts [hunk 1]
```diff
@@ -395,6 +395,23 @@ export function loadSync(
   return createPackageDefinition(root, options!);
 }
 
+export function fromJSON(
+  json: Protobuf.INamespace,
+  root?: Protobuf.Root
```
Comment (murgatroid99) [inline packages/proto-loader/src/index.ts:400]: Protobuf.js objects do not belong in this API. The `root` argument should be removed.

File: packages/proto-loader/src/index.ts [hunk 2]
```diff
@@ -395,6 +395,23 @@ export function loadSync(
...
+): PackageDefinition {
+  const options: Options = json.options || {};
```
Comment (murgatroid99) [inline packages/proto-loader/src/index.ts:402]: Does `json.options` actually contain the same options that are used in `createPackageDefinition`?

File: packages/proto-loader/src/index.ts [hunk 3]
```diff
@@ -395,6 +395,23 @@ export function loadSync(
...
+  const newRoot: Protobuf.Root = root || new Protobuf.Root();
+  if (!!options.includeDirs) {
```
Comment (murgatroid99) [inline packages/proto-loader/src/index.ts:404]: My understanding is that we are not loading any files, so `includeDirs` is not relevant.

**Review by murgatroid99 — CHANGES_REQUESTED** [review]

**Comment by SOVLOOKUP** [comment]
I remove `json.options` because protobufjs has processed it

https://github.com/protobufjs/protobuf.js/blob/7b1efdb8b6f8a677b8770296f750c47763e33aaf/src/root.js#L46

File: packages/proto-loader/src/index.ts [hunk 4]
```diff
@@ -395,6 +395,15 @@ export function loadSync(
   return createPackageDefinition(root, options!);
 }
 
+export function fromJSON(
+  json: Protobuf.INamespace
+): PackageDefinition {
+  const newRoot: Protobuf.Root = new Protobuf.Root();
```
Comment (murgatroid99) [inline packages/proto-loader/src/index.ts:401]: At this point, I don't see any reason to create a new `Root` here. `fromJSON` creates one by default anyway.

File: packages/proto-loader/src/index.ts [hunk 5]
```diff
@@ -395,6 +395,15 @@ export function loadSync(
...
+  const loadedRoot = Protobuf.Root.fromJSON(json, newRoot);
+  loadedRoot.resolveAll();
+  return createPackageDefinition(newRoot, {});
```
Comment (murgatroid99) [inline packages/proto-loader/src/index.ts:404]: The options should be an argument, instead of forcing the default values.

**Review by murgatroid99 — CHANGES_REQUESTED** [review]

File: packages/proto-loader/src/index.ts (hunk 5 above)
Comment (SOVLOOKUP) [reply packages/proto-loader/src/index.ts:404]: Thank you for your reply. This `{}` indeed a bit ugly, but `options` arg is already included in the definition of json description, I think it maybe unnecessary to add additional argument.

File: packages/proto-loader/src/index.ts [hunk 6]
```diff
@@ -395,6 +395,19 @@ export function loadSync(
   return createPackageDefinition(root, options!);
 }
 
+export function fromJSON(
+  json: Protobuf.INamespace,
+  options?: Options
+): PackageDefinition {
+  options = options || {};
+  if (!!options.includeDirs) {
```
Comment (murgatroid99) [inline packages/proto-loader/src/index.ts:403]: Just ignore this option. It doesn't matter.

**Review by murgatroid99 — APPROVED** [review]

**Comment by murgatroid99** [comment]
Thank you for your contribution.

**Pull request merged**